src/cryptsetup/cryptsetup.c

```diff
@@ -992,24 +993,25 @@ static int attach_tcrypt(
                 params.veracrypt_pim = arg_tcrypt_veracrypt_pim;
 
         if (key_data) {
-                params.passphrase = key_data;
-                params.passphrase_size = key_data_size;
+                strv_insert(&tmp_passwords, 0, (char *) key_data);
```

Comment:
lacks error handling (strv_insert() allocates memory and can fail).

Moreover, key_data/key_data_size may contain binary data, including NUL bytes, and is not necessarily suffixed by a NUL byte, hence you cannot use it as a string.